File: packages/shared/src/wire.ts
/**
 * Wire-level constants and types for the roomKit WebSocket gateway.
 *
 * FROZEN — any change here must be coordinated across every lane.
 * See SWARM.md and docs/call-platform-feasibility.md §3-§5.
 */

export const WIRE_VERSION = '0.1.0' as const;

export const AUDIO = {
  sampleRate: 16_000,
  channels: 1,
  bitsPerSample: 16,
  encoding: 'pcm_s16le' as const,
  frameMs: 20,
  samplesPerFrame: 320,
  bytesPerFrame: 640,
} as const;

export const ENDPOINTS = {
  agentWs: (roomId: string, token: string, stream: 'mixed' | 'per-track' = 'mixed') =>
    `/v1/rooms/${roomId}/agent?token=${encodeURIComponent(token)}&stream=${stream}`,
  rooms: '/v1/rooms',
  room: (id: string) => `/v1/rooms/${id}`,
  roomTokens: (id: string) => `/v1/rooms/${id}/tokens`,
  roomTranscript: (id: string) => `/v1/rooms/${id}/transcript`,
  roomSummary: (id: string) => `/v1/rooms/${id}/summary`,
  roomRecording: (id: string) => `/v1/rooms/${id}/recording`,
  webhooks: '/v1/webhooks/livekit',
} as const;

export type AgentJwtClaims = {
  role: 'agent' | 'human';
  identity: string;
  room: string;
  iat?: number;
  exp?: number;
};

export type StreamMode = 'mixed' | 'per-track';

/**
 * Validate a binary audio frame size against the wire contract.
 * Returns true if length is a positive multiple of bytesPerFrame.
 */
export function isValidAudioFrame(byteLength: number): boolean {
  return byteLength > 0 && byteLength % AUDIO.bytesPerFrame === 0;
}

/**
 * Number of 20 ms frames in a buffer.
 */
export function framesIn(byteLength: number): number {
  return Math.floor(byteLength / AUDIO.bytesPerFrame);
}

File: packages/sim-sdk-py/src/callplatform/wire.py
"""Mirror of ``packages/shared/src/wire.ts``.

FROZEN — any change here must be coordinated with the TS source of truth.
The TS file is the canonical specification; this module is a literal Python
transcription. The values are duplicated (not derived) so a regression in
either file is caught by ``tests/test_wire.py``.
"""

from __future__ import annotations

from typing import Callable, Literal, TypedDict

WIRE_VERSION: str = "0.1.0"


class _AudioSpec(TypedDict):
    sampleRate: int
    channels: int
    bitsPerSample: int
    encoding: Literal["pcm_s16le"]
    frameMs: int
    samplesPerFrame: int
    bytesPerFrame: int


# Mirrors `AUDIO` in wire.ts. 16 kHz mono int16 LE, 20 ms = 320 samples = 640 bytes.
AUDIO: _AudioSpec = {
    "sampleRate": 16_000,
    "channels": 1,
    "bitsPerSample": 16,
    "encoding": "pcm_s16le",
    "frameMs": 20,
    "samplesPerFrame": 320,
    "bytesPerFrame": 640,
}


StreamMode = Literal["mixed", "per-track"]


def _agent_ws_path(room_id: str, token: str, stream: StreamMode = "mixed") -> str:
    """Mirror of ``ENDPOINTS.agentWs(roomId, token, stream)`` from wire.ts."""
    from urllib.parse import quote

    return f"/v1/rooms/{room_id}/agent?token={quote(token, safe='')}&stream={stream}"


class _Endpoints(TypedDict):
    agentWs: Callable[..., str]
    rooms: str
    room: Callable[[str], str]
    roomTokens: Callable[[str], str]
    roomTranscript: Callable[[str], str]
    roomSummary: Callable[[str], str]
    roomRecording: Callable[[str], str]
    webhooks: str


ENDPOINTS: _Endpoints = {
    "agentWs": _agent_ws_path,
    "rooms": "/v1/rooms",
    "room": lambda id: f"/v1/rooms/{id}",
    "roomTokens": lambda id: f"/v1/rooms/{id}/tokens",
    "roomTranscript": lambda id: f"/v1/rooms/{id}/transcript",
    "roomSummary": lambda id: f"/v1/rooms/{id}/summary",
    "roomRecording": lambda id: f"/v1/rooms/{id}/recording",
    "webhooks": "/v1/webhooks/livekit",
}


class AgentJwtClaims(TypedDict, total=False):
    role: Literal["agent", "human"]
    identity: str
    room: str
    iat: int
    exp: int


def is_valid_audio_frame(byte_length: int) -> bool:
    """Return True iff ``byte_length`` is a positive multiple of bytesPerFrame.

    Mirrors ``isValidAudioFrame(byteLength)`` in wire.ts.
    """
    return byte_length > 0 and byte_length % AUDIO["bytesPerFrame"] == 0


def frames_in(byte_length: int) -> int:
    """Number of 20 ms frames in a buffer. Mirrors ``framesIn`` in wire.ts."""
    return byte_length // AUDIO["bytesPerFrame"]
